small fixes for ping to properly set state in case of failure

wfurt opened the pull request:
First part allows to pass IPv6 ping tests on FreeBSD. It essentially behaves same as OSX.

Second part was more interesting. On FreeBSD, unprivileged user can only send 56 byte ping.
When running tests with size using ping executable it fails with error. 

SendPingCore always  calls Finish() in finally{}. In case of error, calling Send() would also catch exception and call Finish() again.
That triggers Debug.Assert(_status == InProgress, $"Invalid status: {_status}");
This was introduced in #30000.

This change does not make IPv4 tests with size pass, but the test simply fails instead of aborting and dumping core.

wfurt commented:
test failures seems related. I'm looking into it.

stephentoub commented on src/Common/src/System/Net/NetworkInformation/UnixCommandLinePing.cs:
```diff
@@ -18,7 +18,7 @@ internal static class UnixCommandLinePing
         private static readonly string s_discoveredPing4UtilityPath = GetPingUtilityPath(ipv4: true);
         private static readonly string s_discoveredPing6UtilityPath = GetPingUtilityPath(ipv4: false);
-        private static readonly bool s_isOSX = RuntimeInformation.IsOSPlatform(OSPlatform.OSX);
+        private static readonly bool s_isBSD = RuntimeInformation.IsOSPlatform(OSPlatform.OSX) || RuntimeInformation.IsOSPlatform(OSPlatform.Create("FREEBSD"));
```
Why does this use OSPlatform.Create("FREEBSD") rather than OSPlatform.FreeBSD?

wfurt replied:
tests include this file and build as netstandard. OSPlatform.FreeBSD is not available there.

stephentoub commented on src/Common/src/System/Net/NetworkInformation/UnixCommandLinePing.cs:
```diff
@@ -67,9 +67,9 @@ public static string ConstructCommandLine(int packetSize, string address, bool i
+                if (!ipv4 && s_isBSD)
                 {
-                    // OSX uses -h to set hop limit for IPv6
+                    // OSX nad FreeBSD uses -h to set hop limit for IPv6
```
nit: nad => and

stephentoub commented on src/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs:
```diff
@@ -23,18 +23,11 @@ public partial class Ping
-            }
-            finally
-            {
-                Finish();
```
Isn't this call necessary to reset state?

bartonjs replied:
And, if not, SendPingAsyncCore seems like it needs a similar edit.

wfurt commented:
On Unix, SendPingAsyncCore() calls Finish() in finally block. In case of exception, it was called twice @stephentoub What I'm trying to do here, is making SendPingAsyncCore() same between Unix & Windows. 
Is there some good way how to inject faults in UnitTests? I'm trying to figure out if there is any good way how to write good test for the error path.

stephentoub commented:
> In case of exception, it was called twice 

I see now.  The existing logic here seems unnecessarily complicated, in particular in how things are split between Windows and Unix.  But I agree with the spirit of your change, that either we should only be calling Finish in Ping.cs and not in Ping.Windows/Unix.cs, or we shouldn't be calling Finish in Ping.cs at all, and only doing so in Ping.Windows/Unix.cs.  You went with the latter.  I'd generally prefer the former, as it consolidates more of the logic and reduces the chances of bugs, but that would also be a more significant change, so I'm ok with the path you've taken.

> Is there some good way how to inject faults

Not really.  Some of the networking libraries have UnitTests in addition to FunctionalTests (Ping currently only has the latter), where the UnitTests differ by actually building some of the product source files into the tests, and then using that either as a way to test internal surface area or, more apt for this case, to mock by substituting a custom test implementation for this or that method.  I've not looked deeply enough here to see whether that would be relevant, but maybe it would help.

The only other solution I can think of for automated tests would be to find endpoints known to result in failure, and use those in tests.

wfurt commented:
have more common code makes sense. I'll update the change and I'll do more testing.

wfurt commented:
I made changes to Ping.Unix.cs to match Windows implementation. Rest is just white space cleanup. I can take it if preferable but I decided to keep it since it was already in previous changes.

stephentoub commented on src/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs:
```diff
@@ -23,39 +23,28 @@ public partial class Ping
-            }
-            finally
+            PingReply reply = await t.ConfigureAwait(false);
+            Finish();
```
What happens in cases like:
https://github.com/dotnet/corefx/blob/3172386e3e6f268b1811a831c6890665c3154f8a/src/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs#L345
or
https://github.com/dotnet/corefx/blob/3172386e3e6f268b1811a831c6890665c3154f8a/src/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs#L270
?  Won't we not end up calling Finish?

wfurt replied:
The expectation was that that would be handled in Exception case in Ping.cs. That works for synchronous ping but with Async the exception branch does not get hit as the exception happens on worker thread and bubbles up when called calls await.

wfurt commented:
I did bigger cleanup than I originally anticipated.
- The Unix/Window parts do not call Finish() ever
- I moved some CheckStart() to function where Finish() is handled to make it easier to see the logic
- Functions on Ping.cs are responsible for calling Finish() in all cases - regardless of success or failure. 

On mY system I replaced ping with following fragment:
``` shell
#!/bin/sh
echo foo
exit 5
```

I verified that we get exactly one Finish() call for each attempt to ping with both synchronous and asynchronous version. I also move that away and caused PNSP exception @stephentoub mentioned above and I did same verification.

wfurt commented:
@dotnet-bot test windows x64 debug
@dotnet-bot test windows x86 release

wfurt commented:
@dotnet-bot test windows x64 debug build please
@dotnet-bot test windows x86 release build please

stephentoub commented on src/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs:
```diff
@@ -332,9 +339,12 @@ public Task<PingReply> SendPingAsync(IPAddress address, int timeout, byte[] buff
+            finally
+            {
+                Finish();
+            }
```
@wfurt, how are you validating this change?

This doesn't look right to me.  SendPingAsync calls CheckStart to set the bit, then calls SendPingAsyncCore, which kicks off the async operation, and immediately returns a Task, at which point this code calls Finish even though the operation is still likely in flight.

wfurt commented:
This also contributes to #35351 and #35421. It does not fixes the underlying issue but I verified that if ping fails, we simply  fail the test instead of hitting assert and aborting test suite.

karelz commented:
@wfurt what is missing in the PR to finalize it? (it's 3.5 months old :( and I'd like to clean up old PRs ;))

wfurt commented:
I'm hoping for at least one more review. Last change is 4 days old.

karelz commented:
@wfurt do you mean another round of code review on your latest changes, or one more person to code review (in that case whom)?

stephentoub commented:
I will review again this week.

stephentoub commented on src/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Windows.cs:
```diff
@@ -31,14 +31,14 @@ public partial class Ping
-            // Since isAsync == false, DoSendPingCore will execute synchronously and return a completed 
+            // Since isAsync == false, DoSendPingCore will execute synchronously and return a completed
             // Task - so no blocking here
             return DoSendPingCore(address, buffer, timeout, options, isAsync: false).Result;
```
Nit: it was preexisting your change, but this `.Result` should be changed to `.GetAwaiter().GetResult()`.  If the operation fails with an exception, `.Result` will end up wrapping the exception in an `AggregateException`, whereas `.GetAwaiter().GetResult()` will propagate the same exception instance (we want the latter).

wfurt replied:
updated.

stephentoub commented on src/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs:
```diff
@@ -211,7 +216,6 @@ public PingReply Send(string hostNameOrAddress, int timeout, byte[] buffer, Ping
             CheckArgs(timeout, buffer, options);
-            CheckStart();
 
             return GetAddressAndSend(hostNameOrAddress, timeout, buffer, options);
```
Nit: any idea why this is separate out into its own method?  This appears to be the only call site.

stephentoub commented on src/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs:
```diff
@@ -374,39 +389,39 @@ public void SendAsyncCancel()
             }
+            finally
+            {
+                    Finish();
```
Nit: indentation

stephentoub commented on src/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs:
```diff
@@ -374,39 +389,39 @@ public void SendAsyncCancel()
         private async Task<PingReply> GetAddressAndSendAsync(string hostNameOrAddress, int timeout, byte[] buffer, PingOptions options)
         {
-            bool requiresFinish = true;
+            CheckStart();
```
The caller is calling CheckDisposed, and then CheckStart is calling CheckDisposed again.  Is that on purpose?

wfurt replied:
I removed CheckDisposed from CheckStart and updated rest of the code. Now CheckStart only does state processing balanced out in Finish()

wfurt commented:
/azp run corefx-outerloop-linux

wfurt commented:
/azp run corefx-outerloop-osx

wfurt commented:
/azp run corefx-outerloop-windows
